public/app/features/explore/state/query.test.ts

```diff
@@ -265,22 +273,202 @@ describe('reducer', () => {
+            addQueryRowAction({
+              exploreId: ExploreId.left,
+              query: { refId: 'B', key: 'mockKey', datasource: { type: 'loki' } },
+              index: 0,
+            })
+          )
+          .thenStateShouldEqual({
+            queries: [
+              { refId: 'A', key: 'initialRow', datasource: { type: 'loki' } },
+              { refId: 'B', key: 'mockKey', datasource: { type: 'loki' } },
+            ],
+            queryKeys: ['initialRow-0', 'mockKey-1'],
+          } as unknown as ExploreItemState);
+      });
+    });
+    describe('addQueryRow with mixed datasources disabled', () => {
+      it('should add query row when there is not yet a row', async () => {
+        const exploreId = ExploreId.left;
+        let dispatch: ThunkDispatch, getState: () => StoreState;
+
+        const store: { dispatch: ThunkDispatch; getState: () => StoreState } = configureStore({
+          ...defaultInitialState,
+          explore: {
+            [exploreId]: {
+              ...defaultInitialState.explore[exploreId],
+              queries: [],
+              datasourceInstance: {
+                query: jest.fn(),
+                getRef: jest.fn(),
+                meta: {
+                  id: 'postgres',
+                },
+              },
+            },
+          },
+        } as unknown as Partial<StoreState>);
+
+        dispatch = store.dispatch;
+        getState = store.getState;
+
+        setupQueryResponse(getState());
+
+        await dispatch(addQueryRow(exploreId, 1));
+
+        expect(getState().explore[exploreId].datasourceInstance?.meta?.id).toBe('postgres');
+        expect(getState().explore[exploreId].queries).toHaveLength(1);
+        expect(getState().explore[exploreId].queryKeys).toEqual(['ds1-0']);
+      });
+
+      it('should add another query row if there are two rows already', async () => {
+        const exploreId = ExploreId.left;
+        let dispatch: ThunkDispatch, getState: () => StoreState;
+
+        const store: { dispatch: ThunkDispatch; getState: () => StoreState } = configureStore({
+          ...defaultInitialState,
+          explore: {
+            [exploreId]: {
+              ...defaultInitialState.explore[exploreId],
+              queries: [
+                {
+                  datasource: { type: 'loki', uid: 'ds3' },
+                  refId: 'C',
+                },
+                {
+                  datasource: { type: 'loki', uid: 'ds4' },
+                  refId: 'D',
+                },
+              ],
+              datasourceInstance: {
+                query: jest.fn(),
+                getRef: jest.fn(),
+                meta: {
+                  id: 'loki',
+                },
+              },
+            },
+          },
+        } as unknown as Partial<StoreState>);
+
+        dispatch = store.dispatch;
+        getState = store.getState;
+
+        setupQueryResponse(getState());
+
+        await dispatch(addQueryRow(exploreId, 1));
```

Comment:
These tests were all added in the same describe block as tests around the reducer, even though I wouldn't consider these testing logic in the reducer. 

Like for the first test, it's dispatching an action and testing that the state was affected accordingly, that's logic that is in the reducer. For tests like this one that call `addQueryRow` and test that logic, I think it would make sense to make another describe block. I think that helps structures this test file for future growth, what do you think?